k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/command/UidSearchCommand.java

```diff
@@ -0,0 +1,409 @@
+
+public class UidSearchCommand extends BaseCommand {
+
+    private ImapFolder folder;
+    private MessageRetrievalListener<ImapMessage> listener;
+
+    private Set<Long> sequenceSet;
+    private List<Map.Entry<String, String>> sequenceRanges;
+    private Set<Long> uidSet;
+    private List<Map.Entry<String, String>> uidRanges;
+    private String queryString;
+    private boolean performFullTextSearch;
+    private Date since;
+    private Set<Flag> requiredFlags;
+    private Set<Flag> forbiddenFlags;
+
+    private UidSearchCommand(Builder builder) {
+        super(builder.tag, builder.commandFactory);
+        this.folder = builder.folder;
+        this.listener = builder.listener;
+        this.sequenceSet = builder.sequenceSet;
+        this.sequenceRanges = builder.sequenceRanges;
+        this.uidSet = builder.uidSet;
+        this.uidRanges = builder.uidRanges;
+        this.queryString = builder.queryString;
+        this.performFullTextSearch = builder.performFullTextSearch;
+        this.since = builder.since;
+        this.requiredFlags = builder.requiredFlags;
+        this.forbiddenFlags = builder.forbiddenFlags;
+    }
+
+    @Override
+    public String createCommandString() {
+
+        StringBuilder builder = new StringBuilder(String.format(Locale.US, "%d UID SEARCH ", tag));
+        addSequenceNumbers(builder);
+        addUids(builder);
+        addQueryString(builder);
+        addSince(builder);
+        addRequiredFlags(builder);
+        addForbiddenFlags(builder);
+        if (builder.lastIndexOf(" ") == builder.length() - 1) {
+            builder.deleteCharAt(builder.length() - 1);
+        }
+        return builder.toString();
+    }
+
+    @Override
+    public List<UidSearchCommand> splitCommand(int lengthLimit) {
+
+        List<UidSearchCommand> commands = new ArrayList<>();
+
+        Set<Long> numberSet = null;
+        List<Map.Entry<String, String>> numberRanges = null;
+        boolean splittingUids = false;
+
+        //Currently all usages search on the basis of either sequence numbers or uids, but not both
+        if (sequenceSet != null || sequenceRanges != null) {
+            numberSet = sequenceSet;
+            numberRanges = sequenceRanges;
+            splittingUids = false;
+        } else if (uidSet != null || uidRanges != null) {
+            numberSet = uidSet;
+            numberRanges = uidRanges;
+            splittingUids = true;
+        }
+
+        if (numberSet != null || numberRanges != null) {
+
+            while ((numberSet != null && !numberSet.isEmpty()) || (numberRanges != null && !numberRanges.isEmpty())) {
+
+                UidSearchCommand.Builder builder = this.newBuilder()
+                        .sequenceSet(null)
+                        .sequenceRanges(null)
+                        .uidSet(null)
+                        .uidRanges(null);
+
+                int length = builder.getLength();
+                while (length < lengthLimit) {
```

Comment:
so this increases the length until it's above the length limit? I'm not sure if that's a problem here, but if that's the intent I guess "threshold" is a better name :)